frame/democracy/src/lib.rs

```diff
@@ -709,39 +700,42 @@ decl_module! {
 		/// Register the preimage for an upcoming proposal. This requires the proposal to be
 		/// in the dispatch queue. No deposit is needed.
 		#[weight = SimpleDispatchInfo::FixedNormal(100_000)]
-		fn note_imminent_preimage(origin,
-			encoded_proposal: Vec<u8>,
-			when: T::BlockNumber,
-			which: u32
-		) {
+		fn note_imminent_preimage(origin, encoded_proposal: Vec<u8>) {
 			let who = ensure_signed(origin)?;
 			let proposal_hash = T::Hashing::hash(&encoded_proposal[..]);
 			ensure!(!<Preimages<T>>::exists(&proposal_hash), "preimage already noted");
-			let queue = <DispatchQueue<T>>::get(when);
-			let item = queue.get(which as usize).and_then(|x| x.as_ref())
-				.ok_or("dispatch queue entry not found")?;
-			ensure!(item.0 == proposal_hash, "dispatch queue entry invalid");
+			let queue = <DispatchQueue<T>>::get();
+			ensure!(queue.iter().any(|item| &item.1 == &proposal_hash), "not imminent");
 
 			let now = <system::Module<T>>::block_number();
-			<Preimages<T>>::insert(proposal_hash, (encoded_proposal, who.clone(), <BalanceOf<T>>::zero(), now));
+			let free = <BalanceOf<T>>::zero();
+			<Preimages<T>>::insert(proposal_hash, (encoded_proposal, who.clone(), free, now));
 
-			Self::deposit_event(RawEvent::PreimageNoted(proposal_hash, who, Zero::zero()));
+			Self::deposit_event(RawEvent::PreimageNoted(proposal_hash, who, free));
 		}
 
 		/// Remove an expired proposal preimage and collect the deposit.
 		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
 		fn reap_preimage(origin, proposal_hash: T::Hash) {
 			let who = ensure_signed(origin)?;
 
-			if let Some((_, old, deposit, then)) = <Preimages<T>>::get(&proposal_hash) {
-				let now = <system::Module<T>>::block_number();
-				if now >= then + T::EnactmentPeriod::get() + T::VotingPeriod::get() {
-					// allowed to claim the deposit.
-					let _ = T::Currency::repatriate_reserved(&old, &who, deposit);
-					<Preimages<T>>::remove(&proposal_hash);
-					Self::deposit_event(RawEvent::PreimageReaped(proposal_hash, old, deposit, who));
-				}
-			}
+			let (_, old, deposit, then) = <Preimages<T>>::get(&proposal_hash).ok_or("not found")?;
+			let now = <system::Module<T>>::block_number();
+			// we always for a wait of at least (voting + enactment) to minimise the chance of a
+			// last minute preimage pre-enactment snatch. if a vote begins with no preimage
+			// registered, then you can be sure that the owner cannot yank it prior to enactment.
+			// if it is already registered, then
+			let (voting, enactment) = (T::VotingPeriod::get(), T::EnactmentPeriod::get());
+			let additional = if who == old { Zero::zero() } else { enactment };
+			ensure!(now >= then + voting + additional, "too early");
+
+			let queue = <DispatchQueue<T>>::get();
+			ensure!(queue.iter().all(|item| &item.1 != &proposal_hash), "imminent");
```

Comment:
```suggestion
			ensure!(!queue.iter().any(|item| &item.1 == &proposal_hash), "imminent");
```

Reply:
Please let me know if this is the level of nit you are interested in these reviews.